src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,597 @@
+    (exprs : Typed.Meta.t Expr.Fixed.t list) : bool =
+  match kind with
+  | Fun_kind.StanLib (name, (_ : bool Fun_kind.suffix), _) -> (
+    match name with
+    | "check_matching_dims" -> false
+    | _ -> (
+        let fun_args = List.map ~f:Expr.Typed.fun_arg exprs in
+        (*Right now we can't handle AD real and data matrix funcs
+           that return a matrix :-/*)
+        let is_args_autodiff_real_data_matrix =
+          (*If there are any autodiffable vars*)
+          List.exists
+            ~f:(fun (x, y) ->
+              match (x, y) with
+              | UnsizedType.AutoDiffable, UnsizedType.UReal -> true
+              | _ -> false )
+            fun_args
+          (*And there are any data matrices*)
+          && List.exists
+               ~f:(fun (x, y) ->
+                 match (x, UnsizedType.is_container y) with
+                 | UnsizedType.DataOnly, true -> true
+                 | _ -> false )
+               fun_args
+          (*And there are no Autodiffable matrices*)
+          && List.exists
+               ~f:(fun (x, y) ->
+                 match (x, UnsizedType.contains_eigen_type y) with
+                 | UnsizedType.AutoDiffable, true -> false
+                 | _ -> true )
+               fun_args
+        in
+        match is_args_autodiff_real_data_matrix with
+        | true -> true
+        | false -> List.exists ~f:query_ad_real_data_matrix_assign exprs ) )
+  | CompilerInternal (Internal_fun.FnMakeArray | FnMakeRowVec) -> true
+  | CompilerInternal (_ : 'a Internal_fun.t) -> false
+  | UserDefined ((_ : string), (_ : bool Fun_kind.suffix)) -> false
+
+(**
+ * Query to find the initial set of objects in statements that cannot be SoA.
+ * This is mostly recursive over expressions and statements, with the exception of 
+ * functions and Assignments.
+ * @param in_loop A boolean to specify the logic of indexing expressions. See
+ *  `query_initial_demotable_expr` for an explanation of the logic.
+ *)
+let rec query_initial_demotable_stmt (in_loop : bool)
+    (Stmt.Fixed.({pattern; _}) :
+      (Expr.Typed.Meta.t, Stmt.Located.Meta.t) Stmt.Fixed.t) :
+    string Set.Poly.t =
+  let query_expr = query_initial_demotable_expr in_loop in
+  match pattern with
+  | Stmt.Fixed.Pattern.Assignment
+      ( ((name : string), (ut : UnsizedType.t), lhs)
+      , (Expr.Fixed.({meta= Expr.Typed.Meta.({type_; adlevel; _}); _}) as rhs)
+      ) ->
+      let check_lhs =
+        let lhs_list =
+          Set.Poly.union_list
+            (List.map
+               ~f:
+                 (Index.apply ~default:Set.Poly.empty ~merge:Set.Poly.union
+                    query_expr)
+               lhs)
+        in
+        match is_uni_eigen_loop_indexing in_loop ut lhs with
+        | true -> Set.Poly.add lhs_list name
+        | false -> lhs_list
+      in
+      let check_rhs = query_expr rhs in
+      let both_sides = Set.Poly.union check_lhs check_rhs in
+      let check_bad_assign =
```

Comment:
I just updated the docs with all of the logic rules. This one is if the lhs is a matrix the rhs cannot be a mix of data matrices and parameter scalars. The problem is that in the math library we always return a `Matrix<var>` (the AoS matrix type) for `f(data matrix, param scalar)` like non-unary functions. I'm gonna rewrite this section to lay out whats going on more. There is another intricate thing here I found where we can actually keep things SoA for instances like 

```stan 
functions {
    matrix okay_reduction(real sum_x, matrix y) {
        return sum_x + y;
    }
}
parameters {
    matrix [5, 10] soa_x;
    matrix [5, 10] aos_x;
}
transformed parameters {
    matrix[5, 10] tp_matrix_from_udf_reduced_soa = okay_reduction(sum(soa_x), aos_x);
}
```

Since `soa_x` is only the sum of `soa_x` is used in the UDF we don't need to put it on the list of parameters that need demoted. So that makes the logic a little more complicated but I'll write all that up